### PR TITLE
Adding wNAF multiplication functionality

### PR DESCRIPTION
**Pull request by ggkitsas**
## Description

Adds wNAF multiplication functionality for `Group` elements. It doesn't change the way scalar multiplication is currently performed but rather provides an option to use wNAF.

Two functions are added:
* `wnaf_table`: generates a lookup table for a given `Group` element and `window` size
* `wnaf_mul`: performs the scalar multiplication using a lookup table and a scalar in NAF (can be obtained by `find_wnaf`)
---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: ec/src/group/wnaf.rs
Comment (ValarDragon) [inline ec/src/group/wnaf.rs:29]: Should we change this method to just output the table, and not take in a table as input. I'm not seeing any benefit from having it be a mutative argument?

File: ec/src/group/wnaf.rs
Comment (ValarDragon) [inline ec/src/group/wnaf.rs:5]: Can this function have some comments specifying the algorithm being used for future readers. e.g. just writing out the expected form of the scalar wnaf, and the equation its computing.

File: test-templates/src/groups.rs
Comment (ValarDragon) [inline test-templates/src/groups.rs:79]: Can we lower the max tested table size, I think building a 2^22 table is way too much computation for CI

File: ec/src/group/wnaf.rs
Comment (ValarDragon) [inline ec/src/group/wnaf.rs:35]: Once this is merged, lets make an issue to add parallelism to this

**Review by ValarDragon — CHANGES_REQUESTED** [review]
Thanks for PR'ing this! LGTM, left some requests for comments & minor changes

File: ec/src/group/wnaf.rs
Comment (ggkitsas) [reply ec/src/group/wnaf.rs]: Makes sense, modified :+1:

**Review by ValarDragon — APPROVED** [review]
LGTM!

Minor note, can a changelog entry be added to ### features.
perhaps:
`- #230 (ark-ec) Add `wnaf_mul` implementation for groups`

File: ec/src/group/wnaf.rs
Comment (jon-chuang) [reply ec/src/group/wnaf.rs:38]: not sure you can do this. Base is mutated sequentially

**Comment by jon-chuang** [comment]
I will be making some PRs from celo-org merge which include a wNAF mul combined with GLV. I think what would be nice would be to have the GLV-compatible curve use wNAF mul with a default window size specified in parameters. 

I think we ought to make wNAF (+GLV if applicable) the default mul method, if it is faster.

File: ec/src/group/wnaf.rs
Comment (ValarDragon) [reply ec/src/group/wnaf.rs:38]: You can, its similar to computing all powers of a generator in parallel. You chunk which multiples are to be computed in each thread, and then just combine them together when building the final window.

Comment (jon-chuang) [reply ec/src/group/wnaf.rs:38]: Fair enough, but typically, w =5 or 6 at most for wNAF

**Comment by Pratyush** [comment]
@jon-chuang how does this PR interact with your upcoming wNAF PR?

**Comment by jon-chuang** [comment]
@Pratyush it can be used as an impl in a different file, which I am for as it separates duties.

File: ec/src/group/wnaf.rs
Comment (jon-chuang) [reply ec/src/group/wnaf.rs:38]: Furthermore, for your idea, where it applies, one is probably still better off doubling sequentially to create a table and then running the exponentiation in parallel on the back of that table

File: ec/src/group/wnaf.rs
Comment (Pratyush) [inline ec/src/group/wnaf.rs:5]:
```suggestion
// Computes scalar multiplication of a preprocessed group element by a scalar in windowed non-adjacent form (wNAF).
```

**Comment by Pratyush** [comment]
Thanks for the PR! This looks mostly good to me, barring the concern @jon-chuang raised about using `Group`; it's better to implement it directly for `ProjectiveCurve`, so that you can take advantage of things like mixed addition.

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:34]:
```suggestion
    let mut table = Vec::with_capacity(1 << (window - 1));
```

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:40]:
```suggestion
        base += &dbl;
```

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:11]:
```suggestion
    let mut found_non_zero = false;
```

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:19]:
```suggestion
            found_non_zero = true;
```

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:22]: Should we change the signature to return an `Option<G>`, and return `None` if the table does not contain enough elements?

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:8]: Should we rename this to `wnaf_mul_with_table`, and add a function `wnaf_mul` that works roughly like:
```rust
pub fn wnaf_mul<G: ProjectiveCurve>(element: &G, scalar_wnaf: &[i64]) -> G {
	// calculate table size
	let window_size = something();
	let table = wnaf_table(*element, window_size);
	wnaf_mul_with_table(&table, scalar_wnaf)
}
```

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:14]:
```suggestion
        if found_non_zero {
```

File: ec/src/wnaf.rs
Comment (ggkitsas) [reply ec/src/wnaf.rs:8]: I don't know any particular way of deciding the window size. Do you have anything specific in mind?

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:3]:
```suggestion
use crate::ProjectiveCurve;
use ark_std::vec::Vec;
```

File: ec/src/wnaf.rs
Comment (Pratyush) [reply ec/src/wnaf.rs:8]: Hmm it seems that right now the `wnaf_mul` passes tests because the output of `BigInteger::find_wnaf` assume a window size of just 2, and so we never access something that's not in the table. However, if we change the table size to 1, the tests should fail because `wnaf_mul` accesses a multiple that's not in the table. 

@ggkitsas Do you mind giving me access to your branch so that I can push some fixes for this?

Comment (ggkitsas) [reply ec/src/wnaf.rs:7]: Ahh I see, @Pratyush invitation sent

**Comment by Pratyush** [comment]
@ggkitsas I pushed some updates that should now enable working with arbitrary window sizes. Please take a look and let me know if it makes sense.

**Comment by ggkitsas** [comment]
@Pratyush LGTM, thanks!

File: ec/src/wnaf.rs
Comment (Pratyush) [inline ec/src/wnaf.rs:49]: @ValarDragon @jon-chuang should we normalize the table here?

Also, should we make `fixed_base_msm` use wNAF also?

Comment (jon-chuang) [reply ec/src/wnaf.rs:49]: I think normalizing the table could be done only for fixedbaseMSM? 

Unless mixed addition is significantly faster than the ordinary projective add? Still, for fixed base MSM, one can potentially use the infrastructure created for the batch ops. 

Is it used for something important? I can't recall.

Comment (Pratyush) [reply ec/src/wnaf.rs:49]: Yeah, I was proposing using mixed addition instead of projective add. It's not always much faster, for example it's not worth it for Edwards curves. Another thing that we can do is to make two different methods, one that uses a projective table, and one that use an affine table. (Maybe we can abstract over these?)

Comment (Pratyush) [reply ec/src/wnaf.rs:49]: Let's revisit this later; there's no need to block this PR on this decision.

**Comment by Pratyush** [comment]
This LGTM, so unless folks would like further changes, I'll merge this today.

**Review by ValarDragon — APPROVED** [review]
LGTM!

**Pull request merged**
